src/platform_impl/android/mod.rs

```diff
@@ -106,22 +112,32 @@ impl<T: 'static> EventLoop<T> {
             match self.first_event.take() {
                 Some(EventSource::Callback) => match ndk_glue::poll_events().unwrap() {
                     Event::WindowCreated => {
-                        call_event_handler!(
-                            event_handler,
-                            self.window_target(),
-                            control_flow,
-                            event::Event::Resumed
-                        );
+                        let native_window_lock = ndk_glue::native_window();
```

Comment:
There are two calls remaining to `ndk_glue::native_window()`. IMO these should use `self.native_window_lock` and panic if it was `None` (implying it was called outside a `Resumed` - `Suspended` "block") instead of trying to grab the window on their own regard.

That might require some refactoring as `fn raw_window_handle()` only has access to  `struct Window` which is currently empty, with `EventLoopWindowTarget` supposedly tying them together. I'm not sure what the desired flow is here, the other platform implementations should be able to help out here.

Reply:
I tried something like that first. My attempt ran into `RwLockReadGuard` being `!Send` and I gave that approach up. Will give it another try though.

Reply:
Ok, I came up with something. Let me know what you think.

Reply:
I would have preferred to keep the entire thing in one place without two mutexes/locks. It looks like other backends like Wayland store modifyable data in `EventLoopWindowTarget` through a `RefCell`:

https://github.com/rust-windowing/winit/blob/86748fbc681ea070bc37e90a569c192555b402cd/src/platform_impl/linux/wayland/event_loop/mod.rs#L50-L51

That's where I'd put `native_window_lock` instead. Perhaps in a similar `*State` struct and/or with a `type` alias because it is getting quite long :sweat_smile: 

Maybe the winit-Android developers have a better overview of where to place (mutable) data shared between the `EventLoop` and `Window`; @msiglreith @dvc94ch?

Reply:
I would prefer that too, but I assumed that `Window` needs to be `Send`, which complicates things. I'll take another look at the other backends and see if I can glean anything more.

Reply:
Desktop platforms seem to make sure `Window` is `Send` and synchronize state with the event loop (?). It's not clear how important this is on Android, but if it is, then the fact that you can't put `ReadLockReadGuard` behind an `Arc<Mutex<T>>` becomes a problem.
https://github.com/rust-windowing/winit/blob/86748fbc681ea070bc37e90a569c192555b402cd/src/platform_impl/linux/wayland/window/mod.rs#L46 https://github.com/rust-windowing/winit/blob/86748fbc681ea070bc37e90a569c192555b402cd/src/platform_impl/linux/x11/window.rs#L105

Reply:
@MarijnS95 Still no idea how to solve this without extra synchronization. Some input from a maintainer might be useful.

Reply:
Unfortunately I have not had the time to look at any alternatives either, seems like wrapping our state (either the readlock or raw window handle) inside yet another mutex is the only option - given `Send` limits this might only be possible with the window handle which has to be unconditionally retrieved and stored like this PR does currently.

It would have also been nice if the lock could be transposed to remove the `Option` (only want to hold the lock if it contains a valid window) but that is not possible either.